content/design-decisions/0002-infrastructure-integration-testing.md

```diff
@@ -0,0 +1,90 @@
+
+For an example, let us consider the wildly popular blogging platform of Wordpress.Much of its success was derived from the fact that it made it very easy
+for anyone to get started without needing to be a rocketscientist. We should embrace some of those same elements.  
+
+This is also a terraform-centric approach. It does not address our need to test kops, helm, helmfile, etc. We would prefer not to need to develop (5) new
+techniques for each tool because that will increase the burden on our contributors to master testing frameworks.
+
+### Bats
+
+Bats stands for the "Bash Automated Testing System". It has been around for many years. While the original maintainers have abandoned the project, there is an active community effort and [official fork](https://github.com/bats-core/bats-core) that is committed to maintaining it (as of 2017-09). 
+
+Bats presents a familiar testing DSL for running adhoc commands using the `bash` shell. This is really nice since it is a general way to test commands 
+and outcomes in a largely language/framework agnostic manner. That is, it is not opinionated in the same way as more language-specific testing frameworks. 
+Since the tests are just commands wrapped with a simple DSL plus some syntatic sugar, it is easy for anyone who has ever written a shell script to pick it up
+in a matter of minutes.
+
+The downside with `bats` is that it is written in `bash`, which is not a *modern* programming language. The underlying code that powers `bats` is complicated
+for the uniniated shell programmers. Also, since it is interpreted (not compiled), lots of files/libraries get installed on the system much like a python or ruby app. The tests we write will most likely run commands like `curl` or `jq`, so there are more *native dependencies* that need to be installed. 
+
+Using the `bats` strategy will probably necessistate running tests in a container to avoid needing to install too many native dependecies. 
+
+We noticed that [HashiCorp](https://www.hashicorp.com/) is using `bats-core` for the automated testing of their [`consul-helm`](https://github.com/hashicorp/consul-helm) chart. This is a good example of the verstatility of `bats` for testing.
+
+
+### Makefile
+
+We could of course use a simple (or not so simple) `Makefile` to run our tests. By default `make` will abort on the first non-zero exit codes recieved by any of the commands. 
+
+The problem with using a general purpose tool like `make` is that it is overloading the purpose of the tool: task running. We would need to develop
+our own testing framework and methodology on top of it, which is not our core competency or business differentiator. 
+
+Also, `make` is not a well understood language by developers because a `Makefile` is basically templated shell scrits, so the execution flow is awkward. 
```

Comment:
```suggestion
Also, `make` is not a well understood language by developers because a `Makefile` is basically templated shell scripts, so the execution flow is awkward. 
```